File: src/main/java/h5chart/Vumeter.java
/*******************************************************************************
 * File: Vumeter.java 
 * Copyright (C) 2009-2011 
 * Author: L.D.Nigro 
 * Contributor(s):
 * 
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package h5chart;

import java.util.ArrayList;
import java.util.List;

public class Vumeter extends H5ChartGraph {
	
	public static final String VERTICAL="vertical";
	public static final String HORIZONTAL="horizontal";
	
	/**
	 * 
	 */
	private static final long serialVersionUID = -442728236070719826L;
	
	private String _background= "true";
	
	public boolean isBackground() {
		return new Boolean(_background);
	}

	public void setBackground(boolean background) {
		String an=Boolean.toString(background);
		if(!_background.equals(an)){
			_background = an;
			smartUpdate("background", _background);
		}
	}
	
	
	private double _min= 0 ;

	public double getMin() {
		return _min;
	}

	public void setMin(double min) {
		if(_min!=min){
			_min = min;
			smartUpdate("min", _min);
		}
		
	}

	private int _xpoints=1;
	
	public int getXpoints() {
		return _xpoints;
	}

	public void setXpoints(int points) {
		if(_xpoints!=points){
			_xpoints = points;
			smartUpdate("xpoints", _xpoints);
		}
	}

	private int _ypoints=10;
	
	public int getYpoints() {
		return _ypoints;
	}

	public void setYpoints(int points) {
		if(_ypoints!=points){
			_ypoints = points;
			smartUpdate("ypoints", _ypoints);
		}
	}
			
	private String _orientation=VERTICAL;
	
	public String getOrientation() {
		return _orientation;
	}

	public void setOrientation(String orientation) {
		if(!_orientation.equalsIgnoreCase(orientation)){
			_orientation = orientation;
			smartUpdate("orientation", _orientation);
		}
	}
	
	protected String _currentFont= "bold 12px Arial ";
	
	public String getCurrentFont() {
		return _currentFont;
	}

	public void setCurrentFont(String currentFont) {
		if(!_currentFont.equalsIgnoreCase(currentFont)){
			_currentFont = currentFont;
			smartUpdate("currentFont", _currentFont);
		}
	}

	protected String _currentColor= "grey";

	public String getCurrentColor() {
		return _currentColor;
	}

	public void setCurrentColor(String currentColor) {
		if(!_currentColor.equalsIgnoreCase(currentColor)){
			_currentColor = currentColor;
			smartUpdate("currentColor", _currentColor);
		}
	}
	
	protected String _currentMask= DEFAULT_MASK;

	public String getCurrentMask() {
		return _currentMask;
	}

	public void setCurrentMask(String currentMask) {
		if(!_currentMask.equalsIgnoreCase(currentMask)){
			_currentMask = currentMask;
			smartUpdate("currentMask", _currentMask);
		}
	}
	
	private List ranges;
					
	public Vumeter() {
		ranges=new ArrayList();
	}
	
	public void addRange(Double value, String color){
		ranges.add(new Range(value,color));
	}
	
	public void removeRange(int index){
		ranges.remove(index);
	}
	
	public void clearRanges(){
		ranges.clear();
	}
	
	private String _animate="";
	
	public boolean isAnimate() {
		return new Boolean(_animate);
	}

	public void setAnimate(boolean animate) {
		String an=Boolean.toString(animate);
		if(!_animate.equals(an)){
			_animate = an;
			smartUpdate("animate", _animate);
		}
	}

	private Object _current="";
	
	
	public Object getCurrent() {
		return _current;
	}

	public void setCurrent(String current) {
		if(!_current.equals(current)){
			_current = current;
			smartUpdate("current", current,false);
		}
	}
		
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
			throws java.io.IOException {
		super.renderProperties(renderer);
		
		render(renderer, "min", _min);
		render(renderer, "background", _background);
		render(renderer, "xpoints", _xpoints);
		render(renderer, "ypoints", _ypoints);
		render(renderer, "orientation", _orientation);
		render(renderer, "animate", _animate);
		render(renderer, "current", _current);
		render(renderer, "currentMask", _currentMask);
		render(renderer, "currentFont", _currentFont);
		render(renderer, "currentColor", _currentColor);
		
		String s="";
		for(int n=0;n<ranges.size();n++){
			Range r=(Range)ranges.get(n);
			s+=r.toString()+",";
		}
		s=s.substring(0, s.length()-1);
		s+="";
		
		render(renderer, "ranges", "["+s+"]");

	}

	private class Range{
		Double value;
		String color;
		
		public Range(Double value,String color) {
			this.value=value;
			this.color=color;
		}
		
		public String toString(){
			return "{ value:" + value + ",color: '"  + color + "'}";
		}
	}
}

File: src/main/resources/web/js/h5chart/Vumeter.js
/*******************************************************************************
 * File: Vumeter.js Copyright (C) 2009-2011 Author: L.D.Nigro Contributor(s):
 * 
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

zk.$package('Vumeter');

h5chart.Vumeter = zk.$extends(zk.Widget, {
	_min : 0,
	_background : 'false',
	_orientation : 'vertical',
	_xpoints : 1,
	_ypoints : 10,
	_animate : 'false',
	_current : 0,
	_ranges : new Array(),

	_showTooltip : 'true',
	_shadow : 'true',
	_tooltipFont : '11px Arial',
	_tooltipColor : 'grey',
	_tooltipMask : "function (val){ return val}",
	_currentFont : '11px Arial',
	_currentColor : 'grey',
	_currentMask : "function (val){ return val}",
	_labelFont : '11px Arial',
	_labelColor : 'grey',
	_labelMask : "function (val){ return val}",

	_backImg : null,
	_backImgLeft : 0,
	_backImgTop : 0,
	_backImgWidth : 0,
	_backImgHeight : 0,
	_backId : null,

	isShowTooltip : function() {
		return this._showTooltip;
	},

	setShowTooltip : function(value) {
		this._showTooltip = value;
	},

	isShadow : function() {
		return this._shadow;
	},

	setShadow : function(value) {
		this._shadow = value;
	},

	getTooltipFont : function() {
		return this._tooltipFont;
	},

	setTooltipFont : function(value) {
		this._tooltipFont = value;
	},

	getTooltipColor : function() {
		return this._tooltipColor;
	},

	setTooltipColor : function(value) {
		this._tooltipColor = value;
	},

	getTooltipMask : function() {
		return this._tooltipMask;
	},

	setTooltipMask : function(value) {
		this._tooltipMask = value;
	},

	getCurrentFont : function() {
		return this._currentFont;
	},

	setCurrentFont : function(value) {
		this._currentFont = value;
	},

	getCurrentColor : function() {
		return this._currentColor;
	},

	setCurrentColor : function(value) {
		this._currentColor = value;
	},

	getCurrentMask : function() {
		return this._currentMask;
	},

	setCurrentMask : function(value) {
		this._currentMask = value;
	},

	getLabelFont : function() {
		return this._labelFont;
	},

	setLabelFont : function(value) {
		this._labelFont = value;
	},

	getLabelColor : function() {
		return this._labelColor;
	},

	setLabelColor : function(value) {
		this._labelColor = value;
	},

	getLabelMask : function() {
		return this._labelMask;
	},

	setLabelMask : function(value) {
		this._labelMask = value;
	},

	isBackground : function() {
		return this._ringStyle;
	},

	setBackground : function(value) {
		this._background = value;
	},

	getXpoints : function() {
		return this._xpoints;
	},

	setXpoints : function(value) {
		this._xpoints = value;
	},

	getYpoints : function() {
		return this._ypoints;
	},

	setYpoints : function(value) {
		this._ypoints = value;
	},

	isBackground : function() {
		return this._background;
	},

	setBackground : function(value) {
		this._background = value;
	},

	getOrientation : function() {
		return this._orientation;
	},

	setOrientation : function(value) {
		this._orientation = value;
	},

	getMin : function() {
		return this._min;
	},

	setMin : function(value) {
		this._min = value;
	},

	getCurrent : function() {
		return this._current;
	},

	setCurrent : function(value) {
		eval("this._current =[" + value + "];");
	},

	isAnimate : function() {
		return this._animate;
	},

	setAnimate : function(value) {
		this._animate = value;
	},

	getRanges : function() {
		return this._ranges;
	},

	setRanges : function(value) {
		eval("this._ranges=" + value + ";");
	},

	setPlot : function(value) {
		var o = this.domAttrs_();
		var s = this.uuid;
		var parentId = this.parent.uuid;

		var id = this.uuid;
		var l = new Vumeter();
		var left = parseInt(this.getLeft()) + 2;
		var top = parseInt(this.getTop()) + 2
		var w = parseInt(this.getWidth()) - 7;
		var h = parseInt(this.getHeight()) - 7;

		var ani = false;
		if (this.isAnimate() == 'true')
			ani = true;

		var sh = false;
		if(this.isShadow()=='true')
			sh=true;
		
		var sht = false;
		if(this.isShowTooltip()=='true')
			sht=true;
		
		var back = false;
		if (this.isBackground() == 'true')
			back = true;

		if (!this._backId)
			this._backId = document.getElementById(parentId).getUid();

		l.set( {
			left : left,
			top : top,
			width : w,
			height : h,
			min : this.getMin(),
			background : back,
			orientation : this.getOrientation(),
			regionManager:  this.parent._rm,
			xpoints : this.getXpoints(),
			ypoints : this.getYpoints(),
			current : this.getCurrent(),
			animate : ani,
			uid : this._backId,
			shadow:			sh,
			showTooltip:	sht
		});

		var r = this.getRanges();
		if (r.length > 0) {
			for ( var n = 0; n < r.length; n++) {
				l.serie.add(r[n]);
			}
		} else {
			l.serie.add( {
				value : 0,
				color : 'grey'
			});
		}

		eval("var tooltipmask="+this._tooltipMask);
		l.setTooltipStyle({
			mask : 			tooltipmask ,
			font : 			this.getTooltipFont(),
			color: 			this.getTooltipColor()
		});


		eval("var labelmask="+this._labelMask);
		l.setLabelStyle({
			mask : 			labelmask ,
			font : 			this.getLabelFont(),
			color: 			this.getLabelColor()
		});

			
		eval("var currentmask="+this._currentMask);
		l.setCurrentStyle({
			mask : 			currentmask,
			font : 			this.getCurrentFont(),
			color: 			this.getCurrentColor()
		});

		
		if (this._backImg) {
			this._backImg.putImageData(parentId, this._backImgLeft,
					this._backImgTop);
		}

		this._backImg = new h5Image();
		this._backImg.getImageData(parentId, Math.max(0,left - 2), Math.max(0,top - 2), w + 10, h + 10);
		this._backImgTop = Math.max(0,top - 2);
		this._backImgLeft = Math.max(0,left - 2);
		this._backImgWidth = w + 10;
		this._backImgHeight = h + 10;

		l.draw(parentId);

	},

	detach: function() {
		var parentId=this.parent.uuid;
		this.$supers('detach', arguments);
		if(this._backImg){
			this._backImg.putImageData(parentId, this._backImgLeft, this._backImgTop);
		}
		
		document.getElementById(parentId).updateImage();
		var rm=document.getElementById(parentId).getRegionManager()
		if(rm){
			rm.removeById(this._backId);
		}
		this._backImg= null;
		
	} ,
	
	bind_ : function(desktop, skipper, after) {
		this.$supers('bind_', arguments);
		this.setPlot(true);
	}
});
